gtsfm/utils/metrics.py

```diff
@@ -256,3 +242,119 @@ def log_sfm_summary() -> None:
+    io_utils.save_json_file(os.path.join(REACT_METRICS_PATH, "frontend_full.json"), metrics_list)
+
+
+def aggregate_frontend_metrics(
+    two_view_reports_dict: Dict[Tuple[int, int], TwoViewEstimationReport], angular_err_threshold_deg: float
+) -> None:
+    """Aggregate the front-end metrics to log summary statistics.
+
+    We define "pose error" as the maximum of the angular errors in rotation and translation, per:
+        SuperGlue, CVPR 2020: https://arxiv.org/pdf/1911.11763.pdf
+        Learning to find good correspondences. CVPR 2018:
+        OA-Net, ICCV 2019:
+        NG-RANSAC, ICCV 2019:
+
+    Args:
+        two_view_report_dict: report containing front-end metrics for each image pair.
+        angular_err_threshold_deg: threshold for classifying angular error metrics as success.
+    """
+    num_image_pairs = len(two_view_reports_dict.keys())
+
+    # all rotational errors in degrees
+    rot3_angular_errors = [report.R_error_deg for report in two_view_reports_dict.values()]
+    trans_angular_errors = [report.U_error_deg for report in two_view_reports_dict.values()]
+    rot3_angular_errors = np.array(rot3_angular_errors, dtype=float)
+    trans_angular_errors = np.array(trans_angular_errors, dtype=float)
+    # count number of rot3 errors which are not None. Should be same in rot3/unit3
+    num_valid_image_pairs = np.count_nonzero(~np.isnan(rot3_angular_errors))
+
+    # compute pose errors by picking the max error from rot3 and unit3 errors
+    pose_errors = np.maximum(rot3_angular_errors, trans_angular_errors)
+
+    # check errors against the threshold
+    success_count_rot3 = np.sum(rot3_angular_errors < angular_err_threshold_deg)
+    success_count_unit3 = np.sum(trans_angular_errors < angular_err_threshold_deg)
+    success_count_pose = np.sum(pose_errors < angular_err_threshold_deg)
+
+    # count image pair entries where inlier ratio w.r.t. GT model == 1.
+    all_correct = np.count_nonzero([report.inlier_ratio_gt_model == 1.0 for report in two_view_reports_dict.values()])
+
+    logger.debug(
+        "[Two view optimizer] [Summary] Rotation success: %d/%d/%d",
+        success_count_rot3,
+        num_valid_image_pairs,
+        num_image_pairs,
+    )
+
+    logger.debug(
+        "[Two view optimizer] [Summary] Translation success: %d/%d/%d",
+        success_count_unit3,
+        num_valid_image_pairs,
+        num_image_pairs,
+    )
+
+    logger.debug(
+        "[Two view optimizer] [Summary] Pose success: %d/%d/%d",
+        success_count_pose,
+        num_valid_image_pairs,
+        num_image_pairs,
+    )
+
+    logger.debug(
+        "[Two view optimizer] [Summary] # Image pairs with 100%% inlier ratio:: %d/%d", all_correct, num_image_pairs
+    )
+
+    frontend_metrics = GtsfmMetricsGroup("frontend_summary", [
+        GtsfmMetric("angular_err_threshold_deg", angular_err_threshold_deg), 
+        GtsfmMetric("num_valid_image_pairs", int(num_valid_image_pairs)), 
```

Comment:
before we were also saving:
```
"num_total_image_pairs": int(num_image_pairs),
```
could we still save that, too?

Reply:
wouldnt that also go in some sort of metadata header?

Reply:
I suppose so. But if we aren’t adding the metadata in this PR, I’d prefer we keep this for now until we add it later, so it doesn’t get lost in the shuffle. It could go in both places, I guess.